### ISSUE TITLE
[JENKINS-35367] EMail-Ext Extended Pipeline Support - Templates

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/email-ext-plugin#1147]
---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/davidvanlaatum.png" title="davidvanlaatum's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=davidvanlaatum">davidvanlaatum</a>, imported from: <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-35367" target="_blank">EMail-Ext Extended Pipeline Support - Templates</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/davidvanlaatum.png" title="davidvanlaatum's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=davidvanlaatum">davidvanlaatum</a>
<li><b>status</b>: Closed
<li><b>priority</b>: Minor
<li><b>component(s)</b>: email-ext-plugin
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2016-09-24T03:05:10+00:00
<li><b>votes</b>: 14
<li><b>watchers</b>: 25
<li><b>imported</b>: 20251211-193728
</ul></i>
</details>
<details><summary><i>1 attachment</i></summary>

- [sendMail.groovy](https://issues.jenkins.io/secure/attachment/34244/sendMail.groovy)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>mvollmer:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-35367?focusedId=267189&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-267189">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Would be nice to have this in Pipeline support. We have many jobs using the same templates. When someone joins or leaves the team I just have to edit the matching template and not every build. Since we started mograting builds to pipelines this is a blocker for some builds.</p>
</pre>
</details>

</ul>
</details>

<p>Would be nice to have this in Pipeline support. We have many jobs using the same templates. When someone joins or leaves the team I just have to edit the matching template and not every build. Since we started mograting builds to pipelines this is a blocker for some builds.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/davidvanlaatum.png" title="davidvanlaatum's avatar" /> davidvanlaatum:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-35367?focusedId=267293&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-267293">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I am well aware <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/wink.png" height="16" width="16" align="absmiddle" alt="" border="0"/> currently waiting on the maintainer of the token-macros plugin to get back to me on a problem I encountered using his beta version that adds pipeline support it seems to work in my test cases just waiting for this problem to be resolved <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=slide_o_mix" class="user-hover" rel="slide_o_mix">slide_o_mix</a> </p>
</pre>
</details>

</ul>
</details>

<p>I am well aware <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/wink.png" height="16" width="16" align="absmiddle" alt="" border="0"/> currently waiting on the maintainer of the token-macros plugin to get back to me on a problem I encountered using his beta version that adds pipeline support it seems to work in my test cases just waiting for this problem to be resolved <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=slide_o_mix" class="user-hover" rel="slide_o_mix">slide_o_mix</a> </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>knut_karlsen:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-35367?focusedId=267732&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-267732">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Do you have any eta on the fix? We do not have any emails sent from failing pipeline jobs at the moment. Do you know if there is another way to send emails within the pipeline framework?</p>
</pre>
</details>

</ul>
</details>

<p>Do you have any eta on the fix? We do not have any emails sent from failing pipeline jobs at the moment. Do you know if there is another way to send emails within the pipeline framework?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/davidvanlaatum.png" title="davidvanlaatum's avatar" /> davidvanlaatum:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-35367?focusedId=267866&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-267866">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>No I don't you can send emails using this just can't use the template functionality. ie you have to build the contents of the email in the groovy script</p>
</pre>
</details>

</ul>
</details>

<p>No I don't you can send emails using this just can't use the template functionality. ie you have to build the contents of the email in the groovy script</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/clausfod.png" title="clausfod's avatar" /> clausfod:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-35367?focusedId=268977&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-268977">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Any news on having the Template support via the Pipeline plugin <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> ?</p>
</pre>
</details>

</ul>
</details>

<p>Any news on having the Template support via the Pipeline plugin <img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/> ?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>budhi003:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-35367?focusedId=269750&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-269750">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi Team, David,</p>

<p>Though email-ext involves lot and may need time, is there any basic level of template support existing currently?</p>

<p>Thanks.</p>
</pre>
</details>

</ul>
</details>

<p>Hi Team, David,</p>

<p>Though email-ext involves lot and may need time, is there any basic level of template support existing currently?</p>

<p>Thanks.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/davidvanlaatum.png" title="davidvanlaatum's avatar" /> davidvanlaatum:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-35367?focusedId=270300&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-270300">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The only option at the moment is to build the contents of the email programmatically in the groovy script. I have sent a email to the token macro plugin's maintainer asking for an update</p>
</pre>
</details>

</ul>
</details>

<p>The only option at the moment is to build the contents of the email programmatically in the groovy script. I have sent a email to the token macro plugin's maintainer asking for an update</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/slide_o_mix.png" title="slide_o_mix's avatar" /> slide_o_mix:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-35367?focusedId=270675&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-270675">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Release has been done</p>
</pre>
</details>

</ul>
</details>

<p>Release has been done</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/scm_issue_link.png" title="scm_issue_link's avatar" /> scm_issue_link:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-35367?focusedId=271164&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-271164">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Code changed in jenkins
User: David van Laatum
Path:
 pom.xml
 src/main/java/hudson/plugins/emailext/AttachmentUtils.java
 src/main/java/hudson/plugins/emailext/EmailExtTemplateAction.java
 src/main/java/hudson/plugins/emailext/EmailExtTemplateActionFactory.java
 src/main/java/hudson/plugins/emailext/EmailRecipientUtils.java
 src/main/java/hudson/plugins/emailext/EmailType.java
 src/main/java/hudson/plugins/emailext/ExtendedEmailPublisher.java
 src/main/java/hudson/plugins/emailext/ExtendedEmailPublisherDescriptor.java
 src/main/java/hudson/plugins/emailext/GroovyScriptPath.java
 src/main/java/hudson/plugins/emailext/plugins/ContentBuilder.java
 src/main/java/hudson/plugins/emailext/plugins/CssInliner.java
 src/main/java/hudson/plugins/emailext/plugins/EmailToken.java
 src/main/java/hudson/plugins/emailext/plugins/EmailTrigger.java
 src/main/java/hudson/plugins/emailext/plugins/EmailTriggerDescriptor.java
 src/main/java/hudson/plugins/emailext/plugins/RecipientProvider.java
 src/main/java/hudson/plugins/emailext/plugins/ZipDataSource.java
 src/main/java/hudson/plugins/emailext/plugins/content/AbstractEvalContent.java
 src/main/java/hudson/plugins/emailext/plugins/content/FailedTestsContent.java
 src/main/java/hudson/plugins/emailext/plugins/content/JellyScriptContent.java
 src/main/java/hudson/plugins/emailext/plugins/content/ScriptContent.java
 src/main/java/hudson/plugins/emailext/plugins/content/ScriptContentBuildWrapper.java
 src/main/java/hudson/plugins/emailext/plugins/content/StaticAnalysisUtilities.java
 src/main/java/hudson/plugins/emailext/plugins/content/TemplateContent.java
 src/main/java/hudson/plugins/emailext/plugins/content/TestCountsContent.java
 src/main/java/hudson/plugins/emailext/plugins/content/TriggerNameContent.java
 src/main/java/hudson/plugins/emailext/plugins/recipients/CulpritsRecipientProvider.java
 src/main/java/hudson/plugins/emailext/plugins/recipients/FailingTestSuspectsRecipientProvider.java
 src/main/java/hudson/plugins/emailext/plugins/recipients/FirstFailingBuildSuspectsRecipientProvider.java
 src/main/java/hudson/plugins/emailext/plugins/recipients/ListRecipientProvider.java
 src/main/java/hudson/plugins/emailext/plugins/recipients/RecipientProviderUtilities.java
 src/main/java/hudson/plugins/emailext/plugins/recipients/RequesterRecipientProvider.java
 src/main/java/hudson/plugins/emailext/plugins/recipients/UpstreamComitterRecipientProvider.java
 src/main/java/hudson/plugins/emailext/plugins/trigger/AbortedTrigger.java
 src/main/java/hudson/plugins/emailext/plugins/trigger/AbstractScriptTrigger.java
 src/main/java/hudson/plugins/emailext/plugins/trigger/AlwaysTrigger.java
 src/main/java/hudson/plugins/emailext/plugins/trigger/BuildingTrigger.java
 src/main/java/hudson/plugins/emailext/plugins/trigger/FailureTrigger.java
 src/main/java/hudson/plugins/emailext/plugins/trigger/FirstFailureTrigger.java
 src/main/java/hudson/plugins/emailext/plugins/trigger/FirstUnstableTrigger.java
 src/main/java/hudson/plugins/emailext/plugins/trigger/FixedTrigger.java
 src/main/java/hudson/plugins/emailext/plugins/trigger/FixedUnhealthyTrigger.java
 src/main/java/hudson/plugins/emailext/plugins/trigger/ImprovementTrigger.java
 src/main/java/hudson/plugins/emailext/plugins/trigger/NotBuiltTrigger.java
 src/main/java/hudson/plugins/emailext/plugins/trigger/NthFailureTrigger.java
 src/main/java/hudson/plugins/emailext/plugins/trigger/PreBuildScriptTrigger.java
 src/main/java/hudson/plugins/emailext/plugins/trigger/PreBuildTrigger.java
 src/main/java/hudson/plugins/emailext/plugins/trigger/RegressionTrigger.java
 src/main/java/hudson/plugins/emailext/plugins/trigger/ScriptTrigger.java
 src/main/java/hudson/plugins/emailext/plugins/trigger/SecondFailureTrigger.java
 src/main/java/hudson/plugins/emailext/plugins/trigger/StatusChangedTrigger.java
 src/main/java/hudson/plugins/emailext/plugins/trigger/StillFailingTrigger.java
 src/main/java/hudson/plugins/emailext/plugins/trigger/StillUnstableTrigger.java
 src/main/java/hudson/plugins/emailext/plugins/trigger/SuccessTrigger.java
 src/main/java/hudson/plugins/emailext/plugins/trigger/UnstableTrigger.java
 src/main/java/hudson/plugins/emailext/watching/EmailExtWatchAction.java
 src/main/java/hudson/plugins/emailext/watching/EmailExtWatchJobProperty.java
 src/test/java/hudson/plugins/emailext/AttachmentUtilsTest.java
 src/test/java/hudson/plugins/emailext/EmailExtStepTest.java
 src/test/java/hudson/plugins/emailext/EmailTypeTest.java
 src/test/java/hudson/plugins/emailext/ExtendedEmailPublisherDescriptorTest.java
 src/test/java/hudson/plugins/emailext/ExtendedEmailPublisherMatrixTest.java
 src/test/java/hudson/plugins/emailext/ExtendedEmailPublisherTest.java
 src/test/java/hudson/plugins/emailext/plugins/ContentBuilderTest.java
 src/test/java/hudson/plugins/emailext/plugins/CssInlinerTest.java
 src/test/java/hudson/plugins/emailext/plugins/OnlyRegressionsTest.java
 src/test/java/hudson/plugins/emailext/plugins/ZipDataSourceTest.java
 src/test/java/hudson/plugins/emailext/plugins/content/FailedTestsContentTest.java
 src/test/java/hudson/plugins/emailext/plugins/content/JellyScriptContentTest.java
 src/test/java/hudson/plugins/emailext/plugins/content/ScriptContentTest.java
 src/test/java/hudson/plugins/emailext/plugins/content/TestCountsContentTest.java
 src/test/java/hudson/plugins/emailext/plugins/content/TriggerNameContentTest.java
 src/test/java/hudson/plugins/emailext/plugins/recipients/DevelopersRecipientProviderTest.java
 src/test/java/hudson/plugins/emailext/plugins/recipients/FailingTestSuspectsRecipientProviderTest.java
 src/test/java/hudson/plugins/emailext/plugins/recipients/FirstFailingBuildSuspectsRecipientProviderTest.java
 src/test/java/hudson/plugins/emailext/plugins/recipients/MockUtilities.java
 src/test/java/hudson/plugins/emailext/plugins/recipients/TestUtilities.java
 src/test/java/hudson/plugins/emailext/plugins/trigger/BuildingTriggerTest.java
 src/test/java/hudson/plugins/emailext/plugins/trigger/FirstFailureTriggerTest.java
 src/test/java/hudson/plugins/emailext/plugins/trigger/SecondFailureTriggerTest.java
 src/test/java/hudson/plugins/emailext/plugins/trigger/StatusChangedTriggerTest.java
 src/test/java/hudson/plugins/emailext/plugins/trigger/XNthFailureTriggerTest.java
 src/test/java/hudson/plugins/emailext/recipients/EmailRecipientUtilsTest.java
<a href="http://jenkins-ci.org/commit/email-ext-plugin/15346e064eebf1b870b01a956ff03f92655562c6" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/email-ext-plugin/15346e064eebf1b870b01a956ff03f92655562c6</a>
Log:
  <span class="error">&#91;Fix JENKINS-35367&#93;</span> Pipeline templates (#140)</p>

<p>new token macro version with pipeline support</p>


</pre>
</details>

</ul>
</details>

<p>Code changed in jenkins<br/>
User: David van Laatum<br/>
Path:<br/>
 pom.xml<br/>
 src/main/java/hudson/plugins/emailext/AttachmentUtils.java<br/>
 src/main/java/hudson/plugins/emailext/EmailExtTemplateAction.java<br/>
 src/main/java/hudson/plugins/emailext/EmailExtTemplateActionFactory.java<br/>
 src/main/java/hudson/plugins/emailext/EmailRecipientUtils.java<br/>
 src/main/java/hudson/plugins/emailext/EmailType.java<br/>
 src/main/java/hudson/plugins/emailext/ExtendedEmailPublisher.java<br/>
 src/main/java/hudson/plugins/emailext/ExtendedEmailPublisherDescriptor.java<br/>
 src/main/java/hudson/plugins/emailext/GroovyScriptPath.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/ContentBuilder.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/CssInliner.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/EmailToken.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/EmailTrigger.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/EmailTriggerDescriptor.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/RecipientProvider.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/ZipDataSource.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/content/AbstractEvalContent.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/content/FailedTestsContent.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/content/JellyScriptContent.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/content/ScriptContent.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/content/ScriptContentBuildWrapper.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/content/StaticAnalysisUtilities.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/content/TemplateContent.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/content/TestCountsContent.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/content/TriggerNameContent.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/recipients/CulpritsRecipientProvider.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/recipients/FailingTestSuspectsRecipientProvider.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/recipients/FirstFailingBuildSuspectsRecipientProvider.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/recipients/ListRecipientProvider.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/recipients/RecipientProviderUtilities.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/recipients/RequesterRecipientProvider.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/recipients/UpstreamComitterRecipientProvider.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/trigger/AbortedTrigger.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/trigger/AbstractScriptTrigger.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/trigger/AlwaysTrigger.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/trigger/BuildingTrigger.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/trigger/FailureTrigger.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/trigger/FirstFailureTrigger.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/trigger/FirstUnstableTrigger.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/trigger/FixedTrigger.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/trigger/FixedUnhealthyTrigger.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/trigger/ImprovementTrigger.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/trigger/NotBuiltTrigger.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/trigger/NthFailureTrigger.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/trigger/PreBuildScriptTrigger.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/trigger/PreBuildTrigger.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/trigger/RegressionTrigger.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/trigger/ScriptTrigger.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/trigger/SecondFailureTrigger.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/trigger/StatusChangedTrigger.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/trigger/StillFailingTrigger.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/trigger/StillUnstableTrigger.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/trigger/SuccessTrigger.java<br/>
 src/main/java/hudson/plugins/emailext/plugins/trigger/UnstableTrigger.java<br/>
 src/main/java/hudson/plugins/emailext/watching/EmailExtWatchAction.java<br/>
 src/main/java/hudson/plugins/emailext/watching/EmailExtWatchJobProperty.java<br/>
 src/test/java/hudson/plugins/emailext/AttachmentUtilsTest.java<br/>
 src/test/java/hudson/plugins/emailext/EmailExtStepTest.java<br/>
 src/test/java/hudson/plugins/emailext/EmailTypeTest.java<br/>
 src/test/java/hudson/plugins/emailext/ExtendedEmailPublisherDescriptorTest.java<br/>
 src/test/java/hudson/plugins/emailext/ExtendedEmailPublisherMatrixTest.java<br/>
 src/test/java/hudson/plugins/emailext/ExtendedEmailPublisherTest.java<br/>
 src/test/java/hudson/plugins/emailext/plugins/ContentBuilderTest.java<br/>
 src/test/java/hudson/plugins/emailext/plugins/CssInlinerTest.java<br/>
 src/test/java/hudson/plugins/emailext/plugins/OnlyRegressionsTest.java<br/>
 src/test/java/hudson/plugins/emailext/plugins/ZipDataSourceTest.java<br/>
 src/test/java/hudson/plugins/emailext/plugins/content/FailedTestsContentTest.java<br/>
 src/test/java/hudson/plugins/emailext/plugins/content/JellyScriptContentTest.java<br/>
 src/test/java/hudson/plugins/emailext/plugins/content/ScriptContentTest.java<br/>
 src/test/java/hudson/plugins/emailext/plugins/content/TestCountsContentTest.java<br/>
 src/test/java/hudson/plugins/emailext/plugins/content/TriggerNameContentTest.java<br/>
 src/test/java/hudson/plugins/emailext/plugins/recipients/DevelopersRecipientProviderTest.java<br/>
 src/test/java/hudson/plugins/emailext/plugins/recipients/FailingTestSuspectsRecipientProviderTest.java<br/>
 src/test/java/hudson/plugins/emailext/plugins/recipients/FirstFailingBuildSuspectsRecipientProviderTest.java<br/>
 src/test/java/hudson/plugins/emailext/plugins/recipients/MockUtilities.java<br/>
 src/test/java/hudson/plugins/emailext/plugins/recipients/TestUtilities.java<br/>
 src/test/java/hudson/plugins/emailext/plugins/trigger/BuildingTriggerTest.java<br/>
 src/test/java/hudson/plugins/emailext/plugins/trigger/FirstFailureTriggerTest.java<br/>
 src/test/java/hudson/plugins/emailext/plugins/trigger/SecondFailureTriggerTest.java<br/>
 src/test/java/hudson/plugins/emailext/plugins/trigger/StatusChangedTriggerTest.java<br/>
 src/test/java/hudson/plugins/emailext/plugins/trigger/XNthFailureTriggerTest.java<br/>
 src/test/java/hudson/plugins/emailext/recipients/EmailRecipientUtilsTest.java<br/>
<a href="http://jenkins-ci.org/commit/email-ext-plugin/15346e064eebf1b870b01a956ff03f92655562c6" class="external-link" target="_blank" rel="nofollow noopener">http://jenkins-ci.org/commit/email-ext-plugin/15346e064eebf1b870b01a956ff03f92655562c6</a><br/>
Log:<br/>
  <span class="error">&#91;Fix <a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/35367">JENKINS-35367</a>&#93;</span> Pipeline templates (#140)</p>

<p>new token macro version with pipeline support</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/davidvanlaatum.png" title="davidvanlaatum's avatar" /> davidvanlaatum:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-35367?focusedId=271166&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-271166">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>version 2.50 release and should be available shortly</p>
</pre>
</details>

</ul>
</details>

<p>version 2.50 release and should be available shortly</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/matthewreiter.jpg" title="matthewreiter's avatar" /> matthewreiter:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-35367?focusedId=271311&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-271311">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Even with version 2.50 the FAILED_TESTS macro does not work in pipeline builds; it gets expanded to "FAILED_TESTS is not supported in this context". I believe the problem is that FailedTestsContent was not updated to override the new signature for evaluate() that takes a Run object as opposed to an AbstractBuild object.</p>
</pre>
</details>

</ul>
</details>

<p>Even with version 2.50 the FAILED_TESTS macro does not work in pipeline builds; it gets expanded to "FAILED_TESTS is not supported in this context". I believe the problem is that FailedTestsContent was not updated to override the new signature for evaluate() that takes a Run object as opposed to an AbstractBuild object.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/matthewreiter.jpg" title="matthewreiter's avatar" /> matthewreiter:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-35367?focusedId=271325&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-271325">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>I've opened <a href="https://issues.jenkins.io/browse/JENKINS-38519" title="EMail-Ext Extended Pipeline Support - Failed Tests" class="issue-link" data-issue-key="JENKINS-38519"><del>JENKINS-38519</del></a> for the issue with FAILED_TESTS because it's not directly related to templates.</p>
</pre>
</details>

</ul>
</details>

<p>I've opened <a href="https://issue-redirect.jenkins.io/issue/38519" title="EMail-Ext Extended Pipeline Support - Failed Tests" class="issue-link" data-issue-key="<a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/38519">JENKINS-38519</a>"><del>JENKINS-38519</del></a> for the issue with FAILED_TESTS because it's not directly related to templates.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/nikk699.png" title="nikk699's avatar" /> nikk699:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-35367?focusedId=272551&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-272551">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Is their an example on how to use tokens?</p>
</pre>
</details>

</ul>
</details>

<p>Is their an example on how to use tokens?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i>stegeman:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-35367?focusedId=272581&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-272581">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Thanks to the developer for making the support for the template<img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/><img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/thumbs_up.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nikk699" class="user-hover" rel="nikk699">nikk699</a>, see attachement <span class="nobr"><a href="https://issues.jenkins.io/secure/attachment/34244/34244_sendMail.groovy" title="sendMail.groovy attached to JENKINS-35367">sendMail.groovy<sup><img class="rendericon" src="https://issues.jenkins.io/images/icons/link_attachment_7.gif" height="7" width="7" align="absmiddle" alt="" border="0"/></sup></a></span> for an example with template and tokens</p>
</pre>
</details>

</ul>
</details>

<p>Thanks to the developer for making the support for the template<img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/smile.png" height="16" width="16" align="absmiddle" alt="" border="0"/><img class="emoticon" src="https://issues.jenkins.io/images/icons/emoticons/thumbs_up.png" height="16" width="16" align="absmiddle" alt="" border="0"/></p>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=nikk699" class="user-hover" rel="nikk699">nikk699</a>, see attachement <span class="nobr"><a href="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/34244/sendMail.groovy" title="sendMail.groovy attached to <a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/35367">JENKINS-35367</a>">sendMail.groovy<sup><img class="rendericon" src="https://issues.jenkins.io/images/icons/link_attachment_7.gif" height="7" width="7" align="absmiddle" alt="" border="0"/></sup></a></span> for an example with template and tokens</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/aheritier.png" title="aheritier's avatar" /> aheritier:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-35367?focusedId=277728&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-277728">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=davidvanlaatum" class="user-hover" rel="davidvanlaatum">davidvanlaatum</a></p>

<p>  It could be interesting to add <span class="nobr"><a href="https://issues.jenkins.io/secure/attachment/34244/34244_sendMail.groovy" title="sendMail.groovy attached to JENKINS-35367">sendMail.groovy<sup><img class="rendericon" src="https://issues.jenkins.io/images/icons/link_attachment_7.gif" height="7" width="7" align="absmiddle" alt="" border="0"/></sup></a></span> in the sample directory and a link in the plugin documentation</p>

<p>WDYT?</p>
</pre>
</details>

</ul>
</details>

<p>Hi <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=davidvanlaatum" class="user-hover" rel="davidvanlaatum">davidvanlaatum</a></p>

<p>  It could be interesting to add <span class="nobr"><a href="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-last/refs/heads/main/attachments/34244/sendMail.groovy" title="sendMail.groovy attached to <a class="jira-plain-text-key" href="https://issue-redirect.jenkins.io/issue/35367">JENKINS-35367</a>">sendMail.groovy<sup><img class="rendericon" src="https://issues.jenkins.io/images/icons/link_attachment_7.gif" height="7" width="7" align="absmiddle" alt="" border="0"/></sup></a></span> in the sample directory and a link in the plugin documentation</p>

<p>WDYT?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/davidvanlaatum.png" title="davidvanlaatum's avatar" /> davidvanlaatum:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-35367?focusedId=277729&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-277729">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>not exactly the best of examples eg the recipient providers could be passed directly to emailext rather than going through emailextrecipients. It also replicates much of the features of emailext </p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">	<span class="code-comment">// Append Culprits when the build is not successfull
</span>	<span class="code-keyword">if</span> (currentBuild.result != <span class="code-quote">"SUCCESS"</span>) {
		to &lt;&lt; emailextrecipients([[$class: <span class="code-quote">'CulpritsRecipientProvider'</span>]])
	}
</pre>
</div></div>
<p>is redundant as CulpritsRecipientProvider only adds culprits to the list if the build failed.
It also doesn't handle the body throwing an exception
In theory this is better (untested)</p>
<div class="code panel" style="border-width: 1px;"><div class="codeContent panelContent">
<pre class="code-java">def call(body) {
    <span class="code-comment">// evaluate the body block, and collect configuration into the object
</span>    def config = [:]
    body.resolveStrategy = Closure.DELEGATE_FIRST
    body.delegate = config
    <span class="code-keyword">try</span> {
        body()
    } <span class="code-keyword">catch</span>(e) {
        currentBuild.result = <span class="code-quote">"FAILURE"</span>;
        <span class="code-keyword">throw</span> e;
    } <span class="code-keyword">finally</span> {
        def subject = config.subject ? config.subject : <span class="code-quote">"${env.JOB_NAME} - Build #${env.BUILD_NUMBER} - ${currentBuild.result}!"</span>
        def content = <span class="code-quote">'${JELLY_SCRIPT,template=<span class="code-quote">"<span class="code-keyword">static</span>-analysis"</span>}'</span>
        <span class="code-comment">// Attach buildlog when the build is not successfull
</span>        def attachLog = (config.attachLog != <span class="code-keyword">null</span>) ? config.attachLog : (currentBuild.result != <span class="code-quote">"SUCCESS"</span>)
        <span class="code-comment">// Send email
</span>        emailext(body: content, mimeType: <span class="code-quote">'text/html'</span>,
                replyTo: <span class="code-quote">'$DEFAULT_REPLYTO'</span>, subject: subject,
                to: config.emailRecipients, attachLog: attachLog, recipientProviders: [[$class: <span class="code-quote">'CulpritsRecipientProvider'</span>], [$class: <span class="code-quote">'RequesterRecipientProvider'</span>]])
    }
}
</pre>
</div></div>
</pre>
</details>

</ul>
</details>

<p>not exactly the best of examples eg the recipient providers could be passed directly to emailext rather than going through emailextrecipients. It also replicates much of the features of emailext </p>

<pre>
	<span class="code-comment">// Append Culprits when the build is not successfull
</span>	<span class="code-keyword">if</span> (currentBuild.result != <span class="code-quote">"SUCCESS"</span>) {
		to << emailextrecipients([[$class: <span class="code-quote">'CulpritsRecipientProvider'</span>]])
	}
</pre>
<p>is redundant as CulpritsRecipientProvider only adds culprits to the list if the build failed.<br/>
It also doesn't handle the body throwing an exception<br/>
In theory this is better (untested)</p>

<pre>
def call(body) {
    <span class="code-comment">// evaluate the body block, and collect configuration into the object
</span>    def config = [:]
    body.resolveStrategy = Closure.DELEGATE_FIRST
    body.delegate = config
    <span class="code-keyword">try</span> {
body()
    } <span class="code-keyword">catch</span>(e) {
currentBuild.result = <span class="code-quote">"FAILURE"</span>;
<span class="code-keyword">throw</span> e;
    } <span class="code-keyword">finally</span> {
def subject = config.subject ? config.subject : <span class="code-quote">"${env.JOB_NAME} - Build #${env.BUILD_NUMBER} - ${currentBuild.result}!"</span>
def content = <span class="code-quote">'${JELLY_SCRIPT,template=<span class="code-quote">"<span class="code-keyword">static</span>-analysis"</span>}'</span>
<span class="code-comment">// Attach buildlog when the build is not successfull
</span>def attachLog = (config.attachLog != <span class="code-keyword">null</span>) ? config.attachLog : (currentBuild.result != <span class="code-quote">"SUCCESS"</span>)
<span class="code-comment">// Send email
</span>emailext(body: content, mimeType: <span class="code-quote">'text/html'</span>,
replyTo: <span class="code-quote">'$DEFAULT_REPLYTO'</span>, subject: subject,
to: config.emailRecipients, attachLog: attachLog, recipientProviders: [[$class: <span class="code-quote">'CulpritsRecipientProvider'</span>], [$class: <span class="code-quote">'RequesterRecipientProvider'</span>]])
    }
}
</pre>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/aheritier.png" title="aheritier's avatar" /> aheritier:</i></summary>

<ul>
<li><i>Original <a class="original-jira-link" href="https://issues.jenkins.io/browse/JENKINS-35367?focusedId=277731&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-277731">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>thanks <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=davidvanlaatum" class="user-hover" rel="davidvanlaatum">davidvanlaatum</a></p>
</pre>
</details>

</ul>
</details>

<p>thanks <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=davidvanlaatum" class="user-hover" rel="davidvanlaatum">davidvanlaatum</a></p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `depends_on` from Jira: <a class="jira-relationship" href="https://issue-redirect.jenkins.io/issue/35368">JENKINS-35368</a>]</i>
